soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py

```diff
@@ -0,0 +1,163 @@
+        Sampling time [s].
+    T_ref_seq : Sequence
+        Torque reference sequence [p.u.].
+    u_km1 : 1 x 3 ndarray of ints
+        Previous three-phase switch position (step k-1).
+    state_space : SimpleNamespace 
+        The state-space model of the system.
+    C : 2 x 4 ndarray of ints
+        Output matrix.
+    sim_data : dict
+        Controller data.
+    """
+
+    def __init__(self, solver, lambda_u, Np, Ts, T_ref):
+        """
+        Initialize an IMMpcCurrCtr instance.
+
+        Parameters
+        ----------
+        solver : solver object
+            Solver for an MPC algorithm.
+        lambda_u : float
+            Weighting factor for the control effort.
+        Np : int
+            Prediction horizon.
+        Ts : float
+            Sampling interval [s].
+        T_ref_seq : Sequence
+            Torque reference sequence [p.u.].
+        """
+        self.lambda_u = lambda_u
+        self.Np = Np
+        self.Ts = Ts
+        self.u_km1 = np.array([0, 0, 0])
+        self.T_ref_seq = T_ref
+        self.state_space = SimpleNamespace()
+        self.solver = solver
+
+        # Output matrix
+        self.C = np.array([[1, 0, 0, 0], [0, 1, 0, 0]])
+
+        self.sim_data = {
+            'u': [],
+            'iS_ref': [],
+            't': [],
+            'T_ref': [],
+        }
+
+    def __call__(self, sys, conv, t):
+        """
+        Perform MPC.
+
+        Parameters
+        ----------
+        sys : system object
+            System model.
+        conv : converter object
+            Converter model.
+        t : float
+            Current time [s].
+
+        Returns
+        -------
+        1 x 3 ndarray of floats
+            three-phase switch position or modulating signals.
+        """
+
+        self.state_space = sys.get_discrete_state_space(conv.v_dc, self.Ts)
+
+        # Get the stator current reference for the current step
+        psiR_mag_ref = np.linalg.norm(np.array([sys.x0[2], sys.x0[3]]))
+        T_ref = self.T_ref_seq(t)
+        T_ref = T_ref[0]
+        is_ref_dq = sys.calc_stator_current(psiR_mag_ref, T_ref)
+
+        # Get the grid-voltage angle and calculate the reference in alpha-beta frame
+        theta = np.arctan2(sys.x[3], sys.x[2])
+        i_ref = dq_2_alpha_beta(is_ref_dq, theta)
+
+        # Predict the current reference over the prediction horizon
+        # Make a rotation matrix
+        Ts_pu = self.Ts * sys.base.w
+        delta_theta = sys.wr * Ts_pu
+        R_ref = np.array([[np.cos(delta_theta), -np.sin(delta_theta)], \
+                          [np.sin(delta_theta), np.cos(delta_theta)]])
+
+        # Predict the reference by rotating the current reference
+        y_ref = np.zeros((self.Np, 2))
+        i_ref_temp = i_ref
+        for k in range(self.Np):
+            y_ref[k, :] = np.dot(R_ref, i_ref_temp)
+            i_ref_temp = y_ref[k, :]
+
+        # Solve the control problem
+        uk = self.solver(sys, conv, self, y_ref)
```

Comment:
Given the above comments, I am thinking of structure like this:

* One file for the outer loop: This file generates the current reference (for machine or grid)
* One file with the FCS-MPC current control problem: 
-->We feed into this file the current reference
--> We generate the reference trajectory in the horizon (as you do here)
--> We call the possible way to solve the problem
* One file for the solver